File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.py
K,N = map(int,input().split())  # K개로 N개의 랜선을만든다.
k = []
for _ in range(K):
  k.append(int(input()))
k = sorted(k)
start = 1
end = sum(k)//N #총 길이를 N으로 나눈것보다 더 작은 길이가 나올 수 밖에 없으므로

while start <= end:
  mid = (start + end)//2
  cnt = sum([x//mid for x in k])
  if cnt < N:
    end = mid-1
  elif cnt >= N:
    start = mid+1
    answer = mid
print(answer)

File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.java
import java.util.*;
import java.io.*;

public class Main {

	static int n,k;
	static long max,answer;
	static long[] graph;
	
	public static void main(String[] args) throws Exception{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st;
		
		st = new StringTokenizer(br.readLine());
		n = Integer.parseInt(st.nextToken());
		k = Integer.parseInt(st.nextToken());
		
		graph = new long[n];
		max = 0;
		for (int i = 0; i < n; i++) {
			graph[i] = Long.parseLong(br.readLine());
			max = Math.max(max, graph[i]);
		}
		
		answer = binarySearch(max);
		System.out.println(answer);
		
	}
	
	private static long binarySearch(long num) {
		long left = 1;
		long right = num;
		
		while (left <= right) {
			long mid = (left + right) /2;
			
			// 랜선 개수 확인
			long cnt = 0;
			for (int i = 0; i < n; i++) {
				cnt += graph[i]/mid;
			}
			
			if (cnt >= k) {
				left = mid+1;
			}else {
				right = mid-1;
			}
		}
		
		
		return left-1;
	}

}
